File: backjoon/dfs_bfs/7569_토마토.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
const input = [];

const solution = ([info, ...rest]) => {
  let answer = 0;
  const [m, n, h] = info.split(" ");
  let temp = [];
  const floorArray = [];
  const tomato = [];

  for (let i = 0; i < rest.length; i++) {
    if (temp.length < +n) {
      temp.push(rest[i].split(" ").map((v) => +v));
    }
    if (temp.length === +n) {
      floorArray.push(temp);
      temp = [];
    }
  }

  const bfs = (tomato, floorArray) => {
    const q = [tomato];
    let count = 0;
    const dx = [1, -1, 0, 0, 0, 0];
    const dy = [0, 0, 1, -1, 0, 0];
    const dh = [0, 0, 0, 0, 1, -1];

    while (q.length > 0) {
      const v = q.shift();
      const temp_q = [];
      for (let i = 0; i < v.length; i++) {
        const [a, b, c] = v[i];
        for (let j = 0; j < 6; j++) {
          z = a + dh[j];
          x = b + dy[j];
          y = c + dx[j];
          if (
            x >= 0 &&
            x < n &&
            y >= 0 &&
            y < m &&
            z >= 0 &&
            z < h &&
            floorArray[z][x][y] === 0
          ) {
            temp_q.push([z, x, y]);
            floorArray[z][x][y] = 1;
          }
        }
      }
      if (temp_q.length > 0) {
        count += 1;
        q.push(temp_q);
      }
    }
    return count;
  };

  for (let i = 0; i < h; i++) {
    for (let j = 0; j < n; j++) {
      for (let k = 0; k < m; k++) {
        if (floorArray[i][j][k] === 1) {
          tomato.push([i, j, k]);
        }
      }
    }
  }
  answer = bfs(tomato, floorArray);

  for (let i = 0; i < h; i++) {
    for (let j = 0; j < n; j++) {
      for (let k = 0; k < m; k++) {
        if (floorArray[i][j][k] === 0) {
          answer = -1;
        }
      }
    }
  }
  return answer;
};

rl.on("line", (answer) => {
  input.push(answer.trim());
}).on("close", () => {
  console.log(solution(input));
});

File: backjoon/dfs_bfs/7569_토마토.py
from collections import deque

n, m, h = map(int, input().split())
arr = []
floor = []
count = 0
tomato = []
for _ in range(m*h):
    data = list(map(int, input().split(" ")))
    floor.append(data)
    if len(floor) == m:
        arr.append(floor)
        floor = []

dx = [1, -1, 0, 0, 0, 0]
dy = [0, 0, 1, -1, 0, 0]
dh = [0, 0, 0, 0, 1, -1]
def bfs(tomato):
    global count
    q = deque([tomato])
    while q:
        v = q.popleft()
        temp_q = []
        for item in v:
            a, b, c = item[0], item[1], item[2]
            for k in range(6):
                z = a + dh[k]
                x = b + dy[k]
                y = c + dx[k]
                if 0<=x<m and 0<=y<n and 0<=z<h and arr[z][x][y] == 0:
                    temp_q.append([z, x, y])
                    arr[z][x][y] = 1
        if len(temp_q) > 0:
            count += 1
            q.append(temp_q)

for i in range(h):
    for j in range(m):
        for k in range(n):
            if arr[i][j][k] == 1:
                tomato.append((i, j, k))
bfs(tomato)
check = True
for i in range(h):
    for j in range(m):
        for k in range(n):
            if arr[i][j][k] == 0:
                check = False
if check:
    print(count)
else:
    print(-1)